### ISSUE TITLE
Folding Option<Unit>[]

### ISSUE THREAD

**Issue by NickDarvey** [louthy/language-ext#170]
Hi Paul,

Thanks for an awesome library---this is now a BCL for us on our projects.

I've got a question about folding `Option<Unit>` (though I'm aware this might just be me looking at it the wrong way too...).

We need to ask a bunch of 'condition' actors to evaluate a message for us and if they pass, then we continue on and invoke some 'action' actors.

At the moment, the `condition.Evaluate(x)` method returns `Task<Option<Unit>>` and so we end up with `Task<Option<Unit>[]>`. I'd like to combine all the results using logic with this kind of behaviour: `none * none = none`, `none * unit = none` and `unit * unit = unit`, but `Unit` isn't IMultiplicable- IMultipliable... thingy. How do I get that behaviour from an `Option<Unit>[]`?

(Or should I just be returning bools from `Evaluate(x)`? I've been enjoying using Option<Unit> as a sort of way to continue the computation (or not), so that might be warping my perspective.)

```C#
    var evaluate = fun((SubscriptionConfigurationState cfg, Message m) =>
                 Task.WhenAll(

                    // Get the condition actor proxies
                    cfg.Conditions.Map(c => _conditions.Create(c))

                    // Ask them to evaluate the letter
                    .Map(c => c.Evaluate(m))

                 )

                // Fold the results
                //      We start off with some(unit) because with no
                //      conditions means we want to continue (i.e. unconditional)
                .Map(r => r.Fold(Some(unit), (s, n) => s * n)));

```

Thanks so much!
Nick

**Comment by louthy** [comment]
> Thanks for an awesome library---this is now a BCL for us on our projects.

:)

>  I'd like to combine all the results using logic with this kind of behaviour: none * none = none, none * unit = none and unit * unit = unit, but Unit isn't IMultiplicable- IMultipliable... thingy. How do I get that behaviour from an Option<Unit>[]?

It's a shame that version 2 of lang-ext isn't ready for you, because this would be in the `MonadPlus` category and you'd be able to `mconcat` them.

The three rules you gave (btw you missed out what happens when `unit * none = ???`, I assume it'll be `none`) seem to imply that you could just use a combination of bind and fold:
```c#
    public static Option<Unit> Evalulate(Option<Unit>[] xs) =>
        Evaluate(Some(unit), xs);

    private static Option<Unit> Evalulate(Unit state, Option<Unit>[] xs) =>
        fold(xs, state, (s, x) => from a in s
                                  from b in x
                                  select b);
```
_Disclaimer: I haven't checked this code works!  Looks OK though._
> I've been enjoying using Option as a sort of way to continue the computation

Yeah that's totally the point, the bound value (`Unit` in this case) isn't the controller of that, so `bool` would make no difference really, it's whether it's `None` or not that matters.

**Comment by louthy** [comment]
@NickDarvey Had a bit of a brain freeze moment, example updated!

**Comment by NickDarvey** [comment]
Thanks for the response Paul! Exceeept... I'm having trouble picking the signature (and it can't be inferred).
`s` is the current state and `x` is the next, right?

```C#
static Option<Unit> Evalulate(Option<Unit>[] xs) =>
            fold<Unit, Option<Unit>>(xs, Some(unit), (s, x) => from _ in s
                                                               from b in x
                                                               select b);
```
tells me that `s` doesn't implement the query pattern. 

As an aside, how do you know which style to pick where? I jump between query syntax, fluent and your Prelude class and it never seems as neat as your samples and tests.

**Comment by louthy** [comment]
@NickDarvey Indeed, there seems to be some weird inference thing going on there.  I fixed it by using the fluent style:
```c#
        static Option<Unit> Evalulate(Option<Unit>[] xs) =>
            xs.Fold(Some(unit), (s, x) => from _ in s
                                          from b in x
                                          select b);
```
> As an aside, how do you know which style to pick where? 
I guess I usually just think "What would communicate this idea best to other programmers".  As much as our jobs is to communicate with a computer about our intent, it's also massively about communicating with other members of the team.

So I try to pick the most elegant looking and easy to read approach.  However it doesn't do to mix styles too much, that can be a problem in its own right.  It's relatively easy for me to pick any style I like in response to an issue on here, but if I jumped around style-wise in the same block of code then it'd be a bit shitty.  But that would also fall foul of the "What would communicate this idea best to other programmers" mantra.
